### PR TITLE
Replace stream attribute filter with allowed keys

### PR DESCRIPTION
**Pull request by jtescher**
## Changes

This updates the stream configuration to be compliant with the spec. See discussion in https://github.com/open-telemetry/opentelemetry-specification/issues/3441

## Merge requirement checklist

* [x] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-rust/blob/main/CONTRIBUTING.md) guidelines followed
* [x] Unit tests added/updated (if applicable)
* [x] Appropriate `CHANGELOG.md` files updated for non-trivial, user-facing changes
* [x] Changes in public API reviewed (if applicable)

### CODE REVIEW

File: opentelemetry-sdk/src/metrics/instrument.rs [hunk 1]
```diff
@@ -169,12 +169,15 @@ pub struct Stream {
...
     /// Aggregation the stream uses for an instrument.
     pub aggregation: Option<Aggregation>,
-    /// applied to all attributes recorded for an instrument.
-    pub attribute_filter: Option<Filter>,
+    /// An allow-list of attribute keys that will be preserved for the stream.
+    ///
+    /// Any attribute recorded for the stream with a key not in this set will be
+    /// dropped.
+    ///
+    /// If this set is empty or `None`, all attributes will be kept.
```
Comment (cijothomas) [inline opentelemetry-sdk/src/metrics/instrument.rs:177]: what would be the way to drop every attribute?

Comment (jtescher) [reply opentelemetry-sdk/src/metrics/instrument.rs:177]: hm good point, maybe empty set should be drop all

**Review by cijothomas — APPROVED** [review]

**Comment by cijothomas** [comment]
> ## Changes
> This updates the stream configuration to be compliant with the spec. See discussion in [open-telemetry/opentelemetry-specification#3441](https://github.com/open-telemetry/opentelemetry-specification/issues/3441)
> 
> ## Merge requirement checklist
> * [x]  [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-rust/blob/main/CONTRIBUTING.md) guidelines followed
> * [x]  Unit tests added/updated (if applicable)
> * [x]  Appropriate `CHANGELOG.md` files updated for non-trivial, user-facing changes
> * [x]  Changes in public API reviewed (if applicable)

The PR is simply implementing what is already part of stable spec, and hence the mention of 3441 is irrelevant here. Any reason you linked it?

**Comment by jtescher** [comment]
> The PR is simply implementing what is already part of stable spec, and hence the mention of 3441 is irrelevant here. Any reason you linked it?

was just what I was reading when I noticed it.

**Review by TommyCpp — APPROVED** [review]

**Review by lzchen — APPROVED** [review]

**Pull request merged**